dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,141 @@
+﻿#include "pch.h"
+#include "PushNotificationManager.h"
+#include "PushNotificationManager.g.cpp"
+#include "PushNotificationCreateChannelResult.h"
+#include <winrt\windows.networking.pushnotifications.h>
+#include <winerror.h>
+#include <algorithm>
+#include <vector>
+#include "PushNotificationChannel.h"
+#include <wil/resource.h>
+
+using namespace winrt::Windows::Networking::PushNotifications;
+using namespace winrt;
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    inline constexpr std::uint32_t c_maxBackoffSeconds{ 960 };
+
+    const HRESULT WNP_E_NOT_CONNECTED = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E8L);
+    const HRESULT WNP_E_RECONNECTING = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E9L);
+    const HRESULT WNP_E_BIND_USER_BUSY = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403FEL);
+
+    bool PushNotificationManager::isChannelRequestRetryable(const hresult& hr)
+    {
+        switch (hr)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING:
+        case WNP_E_BIND_USER_BUSY:
+        case RPC_S_SERVER_UNAVAILABLE:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    inline bool IsPackagedProcess()
+    {
+        uint32_t packageFullNameLength = 0;
+        const auto rc{ GetCurrentPackageFullName(&packageFullNameLength, nullptr) };
+
+        if (rc != ERROR_INSUFFICIENT_BUFFER)
+        {
+            winrt::throw_hresult(rc);
+        }
+        return rc == ERROR_INSUFFICIENT_BUFFER;
+    }
+
+    Windows::Foundation::IAsyncOperationWithProgress<Microsoft::ProjectReunion::PushNotificationCreateChannelResult, Microsoft::ProjectReunion::PushNotificationCreateChannelStatus> PushNotificationManager::CreateChannelAsync(winrt::guid remoteId)
+    {
+        static std::vector<winrt::guid> s_remoteIdList;
+        static wil::critical_section s_lock;
+
+        // NOTE: API supports channel requests only for packaged applications
+        IsPackagedProcess();
+
+        check_pointer(&remoteId);
+
+        // Allow to register the progress and complete handler
+        co_await resume_background();
```

Comment:
Where is resume_background defined? What does it does?

Reply:
https://docs.microsoft.com/en-us/uwp/cpp-ref-for-winrt/resume-background - it does two things here.
1. It returns control to the caller allowing the caller to register progress and completion handler
2. It continues with the execution of the API immediately on a thread pool thread.